less_users/tests/test_user_challenge.py

```diff
@@ -1,18 +1,97 @@
 
+from freezegun import freeze_time
+from datetime import date
 
-def test_user_challenge_correct_creation(user, challenge_1_day, create_user_challenge):
-    user_challenge = create_user_challenge
+from less_users.models import UserChallenge, Log
+
+
+def test_user_challenge_correct_creation(
+    user, challenge_1_day, create_user_challenge_day
+):
+    user_challenge = create_user_challenge_day
 
     assert user_challenge in UserChallenge.objects.all()
     assert user_challenge.user == user
     assert user_challenge.challenge == challenge_1_day
 
 
-def test_user_challenge_deactivation(create_user_challenge):
-    user_challenge = create_user_challenge
-    user_challenge.is_active = False
+@pytest.mark.django_db
+@freeze_time("2023-01-01")
+def test_user_challenge_deactivation_when_duration_passed(user, challenge_1_day):
+    new_challenge = UserChallenge(
+        user=user,
+        challenge=challenge_1_day,
+    )
+    new_challenge.save()
+    new_challenge.check_if_active(date_today=date(2023, 1, 3))
+
+    assert not new_challenge.is_active
+
+
+TEST_DIV_DATA = (
+    ("2022-12-01", date(2022, 12, 28), 3),
+    ("2022-12-31", date(2023, 1, 18), 12),
+    ("2023-01-29", date(2023, 3, 1), -1),
+    (date.today().strftime("%Y-%m-%d"), date.today(), 30),
+)
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize("a, b, expected", TEST_DIV_DATA)
+@pytest.mark.freeze_time
+def test_user_challenge_model_days_left_3(
+    user, challenge_3_month, a, b, expected, freezer
+):
+    freezer.move_to(a)
+    new_challenge = UserChallenge(user=user, challenge=challenge_3_month)
+    new_challenge.save()
+
+    date_today = b
+    assert new_challenge.days_left(date_today=date_today) == expected
+
+
+@pytest.mark.django_db
+def test_user_challenge_model_days_left_1(user, challenge_1_day):
+    new_challenge = UserChallenge(user=user, challenge=challenge_1_day)
+    new_challenge.save()
+    new_challenge.check_if_active()
+
+    assert new_challenge.days_left() == 1
+    assert new_challenge.is_active
+
+
+@pytest.mark.django_db
+def test_user_challenge_get_points_for_new_challenge_day(
+    user, create_user_challenge_day
+):
+    assert (
+        create_user_challenge_day.challenge.points
+        == create_user_challenge_day.get_points()
+    )
+
+
+@pytest.mark.django_db
+@freeze_time("2023-01-01")
+def test_user_challenge_get_points_for_new_challenge_month_after_8_days(
+    user, create_user_challenge_month
+):
+    Log.objects.create(user_challenge=create_user_challenge_month)
+
+    assert (
+        create_user_challenge_month.challenge.points
+        == create_user_challenge_month.get_points(date_today=date(2023, 1, 9))
+    )
 
-    assert user_challenge.is_active != True
 
+@pytest.mark.django_db
+@freeze_time("2023-01-01")
+def test_user_challenge_do_not_get_points_for_new_challenge_month_after_5_days(
+    user, create_user_challenge_month
+):
+    Log.objects.create(user_challenge=create_user_challenge_month)
 
+    assert (
```

Comment:
Nie rozumiem tego asserta

Reply:
zmieniłam ten i poprzedni test; faktycznie nic nie wnosiły, ale teraz chyba już widać co mają sprawdzać...?

Reply:
- `test_user_challenge_do_not_get_points_for_new_challenge_month_after_5_days`
ten daje radę;

- `test_user_challenge_get_points_for_new_challenge_month_after_8_days`
porównujesz wartość z `get_points` do ilości punktów z loga? jak to się ma do logiki `get_points`?

- `test_user_challenge_get_points_for_new_challenge_day`
tutaj nawet nie sprawdzasz wyniku `get_points`, tylko porównujesz dwa obiekty, które sama wcześniej stworzyłaś.


Porównaj to z tym:
```python
def test_get_points_when_not_active_returns_0(user, challenge):
    assert challenge.duration == 8
    user_challenge = UserChallenge(
        user=user, challenge=challenge, start_date=datetime(2023, 2, 1)
    )

    result = user_challenge.get_points(date(2023, 10, 13))
    assert result == 0


def test_get_points_when_no_logs_returns_0(user, challenge):
    assert challenge.duration == 8
    user_challenge = UserChallenge(
        user=user, challenge=challenge, start_date=datetime(2023, 2, 1)
    )

    assert not user_challenge.log_set.exists()

    result = user_challenge.get_points(date(2023, 2, 3))
    assert result == 0

    assert user_challenge.is_active


def gets_get_points_when_timedelta_is_less_than_frequency_returns_0(user, challenge):
    assert challenge.duration == 8
    assert challenge.frequency == 2

    user_challenge = UserChallenge.objects.create(
        user=user, challenge=challenge, start_date=datetime(2023, 2, 1)
    )
    log = Log.objects.create(user_challenge=user_challenge, date=datetime(2023, 2, 2))

    assert user_challenge.log_set.exists()

    result = user_challenge.get_points(date(2023, 2, 3))
    assert result == 0


def gets_get_points_when_timedelta_is_more_than_frequency_returns(user, challenge):
    assert challenge.duration == 8
    assert challenge.frequency == 2

    user_challenge = UserChallenge.objects.create(
        user=user, challenge=challenge, start_date=datetime(2023, 2, 1)
    )
    log = Log.objects.create(user_challenge=user_challenge, date=datetime(2023, 2, 2))

    assert user_challenge.log_set.exists()

    result = user_challenge.get_points(date(2023, 2, 7))
    assert result == challenge.points

```